pinot-plugins/pinot-file-system/pinot-s3/src/main/java/org/apache/pinot/plugin/filesystem/S3PinotFS.java

```diff
@@ -0,0 +1,485 @@
+            return true;
+        }
+        if (!isDirectory(srcUri)) {
+            delete(dstUri, true);
+            return copyFile(srcUri, dstUri);
+        }
+        dstUri = normalizeToDirectoryUri(dstUri);
+        ImmutableList.Builder<URI> builder = ImmutableList.builder();
+        Path srcPath = Paths.get(srcUri.getPath());
+        try {
+            boolean copySucceeded = true;
+            for (String directoryEntry : listFiles(srcUri, true)) {
+                URI src = new URI(srcUri.getScheme(), srcUri.getHost(), directoryEntry, null);
+                String relativeSrcPath = srcPath.relativize(Paths.get(directoryEntry)).toString();
+                String dstPath = dstUri.resolve(relativeSrcPath).getPath();
+                URI dst = new URI(dstUri.getScheme(), dstUri.getHost(), dstPath, null);
+                copySucceeded &= copyFile(src, dst);
+            }
+            return copySucceeded;
+        } catch (URISyntaxException e) {
+            throw new IOException(e);
+        }
+    }
+
+    @Override
+    public boolean exists(URI fileUri) throws IOException {
+        try {
+            if (isDirectory(fileUri)) {
+                return true;
+            }
+            if (isPathTerminatedByDelimiter(fileUri)) {
+                return false;
+            }
+            return existsFile(fileUri);
+        }catch(NoSuchKeyException e){
+            return false;
+        }
+    }
+
+    @Override
+    public long length(URI fileUri) throws IOException {
+        try {
+            checkState(!isPathTerminatedByDelimiter(fileUri), "URI is a directory");
+            HeadObjectResponse s3ObjectMetadata = getS3ObjectMetadata(fileUri);
+            checkState((s3ObjectMetadata != null), "File '%s' does not exist", fileUri);
+            if(s3ObjectMetadata.contentLength() == null){
+                return 0;
+            }
+            return s3ObjectMetadata.contentLength();
+        } catch (Throwable t) {
+            throw new IOException(t);
+        }
+    }
+
+    @Override
+    public String[] listFiles(URI fileUri, boolean recursive) throws IOException {
+        try {
+            ImmutableList.Builder<String> builder = ImmutableList.builder();
+            String prefix = normalizeToDirectoryPrefix(fileUri);
+            ListObjectsV2Response listObjectsV2Response;
+            ListObjectsV2Request.Builder listObjectsV2RequestBuilder = ListObjectsV2Request.builder()
+                    .bucket(fileUri.getHost())
+                    .prefix(prefix);
+
+            if (recursive) {
+                ListObjectsV2Request listObjectsV2Request = listObjectsV2RequestBuilder.build();
+                listObjectsV2Response = s3Client.listObjectsV2(listObjectsV2Request);
+            } else {
+                ListObjectsV2Request listObjectsV2Request = listObjectsV2RequestBuilder.delimiter(DELIMITER).build();
+                listObjectsV2Response = s3Client.listObjectsV2(listObjectsV2Request);
```

Comment:
nit: move this repetitive line outside if-else blocks.

Reply:
done